### ISSUE TITLE
[ENHANCEMENT] Math feature wishlist

### ISSUE THREAD

**Issue by JeppeKlitgaard** [Zettlr/Zettlr#1966]
Let me first of all add how fantastic Zettlr looks and feels!

I have been playing around with it, as I am considering changing from Typora, which for math still comes out on top of Zettlr (though Typora has its own caveats)

For me, and lots of other STEM users I am sure, the ergonomics and rendering of math is one of the most important features, and I have a few suggestions on how it could possibly be approved.

### Display errors

I am not familiar with KaTeX, but I am sure it has some way of getting a human-readable error on syntax/parsing errors. This would greatly help 'debug' syntax errors in expressions. Typora has some support for this:

![image](https://user-images.githubusercontent.com/1814611/118412529-db931380-b69a-11eb-8057-7439d4eb69cc.png)

Same TeX in Zettlr:

![image](https://user-images.githubusercontent.com/1814611/118412553-f796b500-b69a-11eb-97b3-2890bd9d5b20.png)

### Live preview

Live preview is another thing I think Typora implements really well, where the math is rendered below as you type:

![image](https://user-images.githubusercontent.com/1814611/118412604-30cf2500-b69b-11eb-980d-9cbc7c3c7e0e.png)

Same TeX in Zettlr:

![image](https://user-images.githubusercontent.com/1814611/118412615-404e6e00-b69b-11eb-8a1a-c0c2f9c873d6.png)

### Global macros

See: #1079 

This would be fantastically helpful and would improve the ergonomics of using Zettlr for STEM-type notes vastly.

(_this is also supported in Typora_)

### KaTeX configurations

Add ability to configure KaTeX options using preferences.

Mainly, this would allow users to define global macros that are persistent across documents.

This is something __not__ supported by Typora.

### References to equations

The ability to easily reference equations would also be cool.

This has also been mentioned in #248 and #640

## Conclusion

The reason for mentioning Typora so many times is not meant as an insult to Zettlr - it is a fantastic piece of software and I think it would easily be the most powerful notetaking method for STEM subjects if these are features are implemented!

Further, it seems like these wouldn't be super hard to implement or require sweeping changes to the codebase! Could v2.1 be the "Math Update"?

* Display Errors
* Live preview
* Global macros
* KaTeX/macro configuration
* References to equations

Maybe this issue could be used as a discussion for cool math features that Zettlr could implement?

**Comment by JeppeKlitgaard** [comment]
@tobiasdiez @nathanlesage 

(Bear with me, this might be a worth a discussion in a separate issue)

I have been playing about with Obsidian for a bit and thought the way they did plugins was quite neat. 

Essentially only the 'supercore' functionality is baked into Obsidian itself, with much of the remaining functionality being implemented as 'core plugins'. For example even quite 'core' functionality like the file explorer and search functionality are implemented as 'core plugins'. These come bundled with the Obsidian and can enabled and disabled, with most of them enabled by default.

Community plugins can then be implemented after unchecking a 'THIS IS POTENTIALLY DANGEROUS' box. I assume these use the same API.

Doing it this way ensures that the plugin API is sufficiently well developed since many core functions of Obsidian are implemented using the API. Further, this would allow the plugin API to mature before eventually supporting community plugins, which I think would be a great feature for Zettlr.

Further, this approach would to some extent also help make Zettlr more flexible - from looking at other issues, it seems that some users want the academic, article/writing focused functionality that Zettlr does so well, while others want more note-writing functionality. I think Zettlr could reasonably fit both needs!

I was thinking, maybe the Math functionality of Zettlr should be implemented as a 'core plugin', rather than baked into Zettlr itself?

Anyway, this was just a thought - thank you for making Zettlr open source and well done on making such a nice suite.

**Comment by tobiasdiez** [comment]
Personally I have bad experiences with plugin systems while maintaining JabRef. It makes development harder since you need to preserve the public API and on the other hand you have no guarantee that the developer of the plugins keeps updating them. This may leads to a very chaotic state where you have a couple of half-working half-maintained plugins. In my opinion, you need to have a very active and large developer community to profit from plugins (see VS code for example). For now I don't see any advantage of the plugin vs accepting PRs for Zettlr itself.

That being said, I do think it might make sense to have an option to hide or disable certain features, maybe even some preference preset for "academic writing" vs "note taking" vs "book writing" vs ...

**Comment by furkantaskin** [comment]
Great explanation for math editors and comparison about MathJax and KaTeX. In my opinion, KaTeX is better in user interaction. MathJax improved itself in v3 with performance and usability but what gives us the problem is, you need many extensions to use MathJax in efficient way just KaTeX offers. 

As an example, MathJax v3 lost the support for line breaking option (\\ or \newline). To use newline option, you need to add \displaylines macro. It may look like nothing hard to use but if you want to transfer your export your files as .md extension, some softwares using KaTeX as math editor would not be able to properly render the math commands. Additionally, (I am using this macro really much for my matrices) global macros like \gdef are still not native with MathJax. You have to add some extensions to use them.

There is an open issue in MathJax about newline: https://github.com/mathjax/MathJax/issues/2312

Although MathJax is great in performance and rendering optimization against KaTeX, newline macro is still the biggest problem in MathJax for many users. 


> MathJax 3 supports the `\require{...}` command which can dynamically load packages. User-defined macros can also be defined dynamically using standard LaTeX commands.

MathJax is great with extensions and support for various macros but still as I said, these options are making it hard to moving files to another softwares or using them in different notes


> I am not familiar with KaTeX, but I am sure it has some way of getting a human-readable error on syntax/parsing errors. This would greatly help 'debug' syntax errors in expressions. Typora has some support for this (Typora uses MathJax 2.7):

This would be the great step for KaTeX optimization I guess. Sometimes, finding error in long-written KaTeX format becomes a great problem.

KaTeX is a great package but following the major updates makes it better. And as I see in debug mode, Zettlr is currently using previous major version for KaTeX.

**Comment by nathanlesage** [comment]
@tobiasdiez If you want to prevent stalebot from touching issues, you can add the `pinned` label. I've whitelisted both that and `bug` labels from the stalebot.

**Comment by mcaruel** [comment]
> Let me first of all add how fantastic Zettlr looks and feels!
> 
> I have been playing around with it, as I am considering changing from Typora, which for math still comes out on top of Zettlr (though Typora has its own caveats)
> 
> For me, and lots of other STEM users I am sure, the ergonomics and rendering of math is one of the most important features, and I have a few suggestions on how it could possibly be approved.
> 
> I figure this issue could be a good overview of the state of math in Zettlr.
> 
> ## Suggestions
> ### Display errors
> I am not familiar with KaTeX, but I am sure it has some way of getting a human-readable error on syntax/parsing errors. This would greatly help 'debug' syntax errors in expressions. Typora has some support for this (Typora uses MathJax 2.7):
> 
> ![image](https://user-images.githubusercontent.com/1814611/118412529-db931380-b69a-11eb-8057-7439d4eb69cc.png)
> 
> Same TeX in Zettlr:
> 
> ![image](https://user-images.githubusercontent.com/1814611/118412553-f796b500-b69a-11eb-97b3-2890bd9d5b20.png)
> 
> ### Live preview
> Live preview is another thing I think Typora implements really well, where the math is rendered below as you type:
> 
> ![image](https://user-images.githubusercontent.com/1814611/118412604-30cf2500-b69b-11eb-980d-9cbc7c3c7e0e.png)
> 
> Same TeX in Zettlr:
> 
> ![image](https://user-images.githubusercontent.com/1814611/118412615-404e6e00-b69b-11eb-8a1a-c0c2f9c873d6.png)
> 
> ### Global macros
> See: #1079
> 
> This would be fantastically helpful and would improve the ergonomics of using Zettlr for STEM-type notes vastly.
> 
> (_this is also supported in Typora_)
> 
> ### KaTeX configurations
> Add ability to configure KaTeX options using preferences.
> 
> Mainly, this would allow users to define global macros that are persistent across documents.
> 
> This is something **not** supported by Typora.
> 
> #### MathJax 3 implementation
> MathJax 3 supports the `\require{...}` command which can dynamically load packages. User-defined macros can also be defined dynamically using standard LaTeX commands.
> 
> It would be neat if `preamble.sty` files in the workspace would apply to files at the same level or lower in the file tree. This would allow:
> 
> * Global macros by putting them in `<workspace root>/preamble.sty`
> * Folder specific macros by putting them in `<workspace root>/some_subfolder/preamble.sty` which would only apply to files in `some_subfolder` and below.
> 
> ### References to equations
> The ability to easily reference equations would also be cool.
> 
> This has also been mentioned in #248 and #640
> 
> ### Centered equations
> Double $$ math blocks should be centered.
> 
> See: #648
> 
> This is supported in Typora.
> 
> ### Option to automatically number equations
> Automatic numbering of equations could be a neat feature, and it works well in Typora.
> 
> This should not be in place of support for proper referencing of equations described earlier
> 
> ## MathJax 3?
> Mathjax 2.7 is awfully slow, which I imagine was one of the deciding reasons for choosing KaTeX over MathJax back when that decision was made for Zettlr.
> 
> This has changed with MathJax 3, however. MathJax is a more mature project, has slightly nicer rendering, and has far more LaTeX features implemented than KaTeX. It has a fairly large repository of official extensions as well (particularly the fairly large `physics` package).
> 
> If you do decide to give the math rendering some (more) love in Zettlr, it might be worthwhile to evaluate the pros and cons of KaTeX and MathJax again.
> 
> See: [Speed and rendering comparison](https://www.intmath.com/cg5/katex-mathjax-comparison.php) See: [MathJax Extensions](https://docs.mathjax.org/en/latest/input/tex/extensions/index.html)
> 
> ## Conclusion
> The reason for mentioning Typora so many times is not meant as an insult to Zettlr - it is a fantastic piece of software and I think it would easily be the most powerful notetaking method for STEM subjects if these are features are implemented!
> 
> Further, it seems like these wouldn't be super hard to implement or require sweeping changes to the codebase! Could v2.1 be the "Math Update"?
> 
> * [ ]  Display Errors
> * [ ]  Live preview
> * [ ]  Global macros
> * [ ]  KaTeX/macro configuration
> * [ ]  References to equations
> * [x]  Centering of equations: Implemented by [Render equations with double-dollar in display mode #648](https://github.com/Zettlr/Zettlr/pull/648).
> * [ ]  Automatic numbering of equations
> 
> Maybe this issue could be used as a discussion for cool math features that Zettlr could implement?
> 
> While I am entirely unfamiliar with TypeScript, I might want to play around with implementing some of these over the summer (though likely I won't have time to get around to it, sadly). I am particularly interested in hearing @nathanlesage opinions on MathJax3 vs KaTeX.

Are there news about the possibility to define user latex command to speed up equation writing. This is the only thing that prevents me from moving to writing directly in LaTeX to writing in .md

**Comment by tobiasdiez** [comment]
I would suggest to subscribe to https://github.com/Zettlr/Zettlr/pull/3099 so that you get informed as soon as there is further progress in this direction.
